Add kwarg to index TPRParser from 0 or 1

lilyminium opened the pull request:
Fixes #2364 

Changes made in this Pull Request:
 - Added keyword argument to index TPRParser from 0 or 1
 - Set the default to index from 1, changing the default behaviour

~This is easily undoable. I read the discussion in #2364 as supporting both including a kwarg and a warning. If that is not consensus, please let me know. I'll update CHANGELOG once we have this nailed down :-)~


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

pep8speaks commented:
Hello @lilyminium! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`package/MDAnalysis/topology/tpr/utils.py`](https://github.com/MDAnalysis/mdanalysis/blob/daf8b27494b26f0219be6e40668845ef75f56e4e/package/MDAnalysis/topology/tpr/utils.py):

> [Line 383:69](https://github.com/MDAnalysis/mdanalysis/blob/daf8b27494b26f0219be6e40668845ef75f56e4e/package/MDAnalysis/topology/tpr/utils.py#L383): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

* In the file [`testsuite/MDAnalysisTests/topology/test_tprparser.py`](https://github.com/MDAnalysis/mdanalysis/blob/daf8b27494b26f0219be6e40668845ef75f56e4e/testsuite/MDAnalysisTests/topology/test_tprparser.py):

> [Line 294:16](https://github.com/MDAnalysis/mdanalysis/blob/daf8b27494b26f0219be6e40668845ef75f56e4e/testsuite/MDAnalysisTests/topology/test_tprparser.py#L294): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment

jbarnoud reviewed (approved):
Looks good to me. I am not sure the deprecation message is very clear, though. I would phrase the message to say that from version 2.0, resid will start at 1 rather than 0.

jbarnoud reviewed (changes requested):
I just realized it is missing an entry in the changelog and a "versionchanged".

lilyminium commented:
Thanks @jbarnoud, I've updated CHANGELOG now. I also realised this PR is to develop, so I removed the deprecation warning and changed the default. #3153 is the "first" addition of this to master with the deprecation warning and the status quo default of ``tpr_resid_from_one=False``.

orbeckst commented on package/MDAnalysis/topology/tpr/utils.py:
```diff
@@ -46,6 +46,8 @@
 
 The module also contains the :func:`do_inputrec` to read the TPR header with.
 """
+import warnings
+
```
I don't think `warnings` is needed for the PR against develop. In 2.0.0 we will index TPR resids from 1 by default.

orbeckst commented on package/CHANGELOG:
```diff
@@ -128,6 +128,8 @@ Enhancements
     explicit in the topology (Issue #2468, PR #2775)
   
 Changes
+  * TPRParser now loads TPR files with `tpr_resid_from_one=True`
+    by default (PR #3152)
```
```suggestion
  * TPRParser now loads TPR files with `tpr_resid_from_one=True` by default, 
    which starts TPR resid indexing from 1 (instead of 0 as in 1.x) (Issue #2364, PR #3152)
```

orbeckst commented on package/MDAnalysis/topology/tpr/utils.py:
```diff
@@ -284,7 +286,7 @@ def extract_box_info(data, fver):
 
 
-def do_mtop(data, fver):
+def do_mtop(data, fver, tpr_resid_from_one=True):
```
I feel the default should remain `False` because that is what the TPR actually contains. IMHO the TPR functions themselves could/should stay as faithful to the original format as possible. Assume someone wrote their own TPR parser with our functions. With the new default you break their code needlessly, with the old defaults nothing changes and our own parser just makes the well-documented choice to index from 1.

orbeckst replied:
I can be convinced otherwise (ie to let default `True` stand) but wanted to discuss this point first before approving.

lilyminium replied:
Others may want to chime in (maybe the most knowledgeable, @jbarnoud) but I agree with your argument.

orbeckst replied:
I am going to approve because it looks sensible from my point but let @jbarnoud also say something before merging.

jbarnoud replied:
TPR indeed index from 0, but it is a bit moot because you do not read a TPR directly. Instead, you end up reading your gro and ITP files to make sense of your TPR so consistency there seems the right thing to do.

jbarnoud replied:
My previous objection was that it would break the existing code, but the deprecation in 1.1.0 should be enough.

orbeckst reviewed (changes requested):
Thank you! I was first confused by the PR description but then saw that this is one against develop.

I have minor improvements and one bigger discussion point regarding the default value inside the low-level TPR reader code.

orbeckst commented:
@lilyminium the issue description is not correct, is it?

> Set the default to index from 0, i.e. status quo

*This* PR (against develop) *does* change the default indexing to 1-based.

orbeckst commented:
Note that anything referring to 1.0.2 should become 1.1.0 (see comment on #3153 ).

orbeckst commented:
@jbarnoud any last comments? Otherwise I'll merge tomorrow.
